src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -13,28 +15,95 @@ public WhenKeywordRecommender()
         {
         }
 
-        protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
+        protected override bool IsValidContext(int position, CSharpSyntaxContext context,
+            CancellationToken cancellationToken)
         {
-            if (context.IsCatchFilterContext)
+            return context.IsCatchFilterContext || IsAfterCompleteExpressionOrPatternInCaseLabel(context, cancellationToken);
+        }
+
+        private static bool IsAfterCompleteExpressionOrPatternInCaseLabel(CSharpSyntaxContext context,
+            CancellationToken cancellationToken)
+        {
+            var switchLabel = GetAncestorUntilStatement<SwitchLabelSyntax>(context.TargetToken);
+            if (switchLabel == null)
             {
-                return true;
+                return false;
             }
 
-            // case int i $$
-            // case int i w$$
-            if (IsIdentifierInCasePattern(context.TargetToken.Parent))
+            var expressionOrPattern = switchLabel.ChildNodes().FirstOrDefault();
+            if (expressionOrPattern == null) // Oh well. It must have been a default label.
             {
-                return true;
+                return false;
+            }
+
+            // If the last token missing, the expression is incomplete - possibly because of missing parentheses,
+            // but not necessarily. We don't want to offer 'when' in those cases. Here are some examples that illustrate this:
+            // case |
+            // case 1 + |
+            // case (1 + 1 |
+            // case new |
+
+            // Also note that if there's a missing token inside the expression, that's fine and we do offer 'when':
+            // case (1 + ) |
+
+            var lastToken = expressionOrPattern.GetLastToken(includeZeroWidth: true);
+            if (lastToken.IsMissing)
+            {
+                return false;
+            }
+
+            // Only offer if we're at the last token of the expression/pattern, not inside it.
+
+            if (lastToken == context.TargetToken)
+            {
+                return NotATypeName(expressionOrPattern);
+            }
+            else if (lastToken == context.LeftToken && expressionOrPattern is DeclarationPatternSyntax declarationPattern)
+            {
+                // case constant w|
+
+                // We might have a partially written 'when' keyword, which causes this to be parsed as a pattern.
+                // lastToken will be 'w' (LeftToken) as opposed to 'constant' (TargetToken). We handle this as a special case.
+
+                return NotATypeName(declarationPattern.Type);
             }
 
             return false;
+
+            bool NotATypeName(SyntaxNode node)
```

Comment:
Naming is hard, especially with negatives... After 10 minutes of pondering and constantly changing the name, I settled on this. Even adding 'Is' makes this slightly less readable, not more. A longer name wouldn't be an improvement here.